chore: Improve the way a npm build script is found.

lholmquist opened the pull request:
* This commit will first install just the production-only dependencies. Then a script
will check the package.json for a build script.  If non is found, it just continues, but
if there is a build script, `npm install` will be called again to install the dev dependencies.
Then `npm run build` is execute.

* This reason for doing this way was so we weren't installing all dependecies if we didn't need them


At first i thought we could use something like [jq](https://stedolan.github.io/jq/) to inspect the package.json, but we would have to install that library. Then i realized we could just use node instead, since it is already installed


This should probably also go into the 10.x and 8.x branches 

connects to #149

uocxp commented:
i would suggest doing the `npm install --only=production` after checking if build exists. Using the current code i am getting errors of missing dependencies when doing a production install.

```
if [ "$HAS_BUILD" == true ]; then
			# Do a npm install to get the dev depdencies
			echo "---> Installing dev dependencies"
			npm install
			#do not fail when there is no build script
			echo "---> Building in production mode"
			npm run build --if-present

			echo "---> Pruning the development dependencies"
			npm prune
else
                       		echo "---> Using 'npm install -s --only=production'"
                                npm install -s --only=production
fi
```

lholmquist commented:
@uocxp if you still have a postinall script in your package.json to run your build script then this is probably why you are gettings errors during the production install.  Do you have a repo you can share so i can see if that is the case?

With this new code in this PR, you wouldn't need that postinstall script to run the build

uocxp commented:
```
{
  "name": "test-app",
  "version": "0.1.0",
  "description": "Intents Routing Communication with the NLU for Intents matching Is responsible for the module chain management and orchestration Fulfillment endpoint, acts as a relay to the fulfillment resolver component",
  "main": "dist/server/server.js",
  "types": "dist/server/server.d.ts",
  "private": true,
  "engines": {
    "node": "<9"
  },
  "scripts": {
    "docs:install": "gitbook install",
    "docs:serve": "gitbook serve",
    "docs:build": "gitbook build",
    "docs:pdf": "gitbook pdf",
    "prettier": "prettier --list-different '**/*.ts'",
    "precommit": "lint-staged",
    "lint": "tslint -c tslint.json -p tsconfig.json",
    "test": "jest",
    "test:ci": "jest --clearCache && jest --ci ",
    "tsc": "tsc",
    "watch-tsc": "tsc -w",
    "clean": "rimraf dist && rimraf logs",
    "watch-node": "nodemon ./dist/server/server.js",
    "start": "node ./dist/server/server.js",
    "build": "tsc",
    "postinstall": "npm run build",
    "debug": "concurrently -k -p \"[{name}]\" -n \"TypeScript,Node\" -c \"yellow.bold,green.bold\" \"npm run watch-tsc\" \"npm run watch-node\"",
    "deploy": "npm run build && rsync -rv --exclude '.git' --exclude 'node_modules' . root@demo.mswag.wor.net:/usr/local/bot-engine",
    "db:dump": "mongodump --db botengine -o demo-data",
    "db:load": "mongorestore demo-data",
    "db:clear": "mongo botengine --eval 'db.dropDatabase()'"
  },
  "husky": {
    "hooks": {
      "pre-commit": "npm run precommit",
      "pre-push": "npm run test"
    }
  },
  "lint-staged": {
    "linters": {
      "*.{ts,tsx}": [
        "npm run lint",
        "prettier --write",
        "git add"
      ]
    }
  },
  "repository": {
    "type": "git",
    "url": "xxxxxxxxxx"
  },
  "keywords": [
    "node"
  ],
  "author": "xxxxxxxx",
  "dependencies": {
    "axios": "^0.18.0",
    "body-parser": "^1.18.3",
    "compare-versions": "^3.4.0",
    "dialogflow": "^0.6.0",
    "express": "^4.16.3",
    "flat": "^4.1.0",
    "mongodb": "^3.1.6",
    "nodemailer": "^4.6.8",
    "rxjs": "^6.3.3",
    "uuid": "^3.3.2",
    "winston": "^3.1.0"
  },
  "devDependencies": {
    "@types/body-parser": "^1.17.0",
    "@types/compare-version": "^0.1.31",
    "@types/express": "^4.16.0",
    "@types/flat": "^0.0.28",
    "@types/jest": "^23.3.2",
    "@types/mongodb": "^3.1.9",
    "@types/mongodb-memory-server": "^1.8.0",
    "@types/moxios": "^0.4.8",
    "@types/node": "^10.11.0",
    "@types/nodemailer": "^4.6.5",
    "@types/nodemailer-stub-transport": "^1.1.4",
    "@types/supertest": "^2.0.6",
    "@types/uuid": "^3.4.4",
    "concurrently": "^4.0.1",
    "gitbook-cli": "^2.3.2",
    "husky": "^1.0.0",
    "jest": "^23.6.0",
    "lint-staged": "^7.3.0",
    "moxios": "^0.4.0",
    "nodemailer-stub-transport": "^1.1.0",
    "nodemon": "^1.18.4",
    "prettier": "^1.14.3",
    "rimraf": "^2.6.2",
    "supertest": "^3.3.0",
    "ts-jest": "^23.10.1",
    "tslint": "^5.11.0",
    "tslint-config-airbnb": "^5.11.0",
    "tslint-config-prettier": "^1.15.0",
    "typescript": "~3.0.3",
    "mongodb-memory-server": "^2.4.3"
  }
}

```

uocxp commented:
yes i have a postinstall script in order to install tsc. when i remove the postinstall script  i get  that `tsc: command not found`

lholmquist commented:
@uocxp i've updated the script with your suggestion.  We don't need to be calling postinstall scripts more than once.

The other reason why it didn't work, is becuase by default the NODE_ENV is set to "production", and we only override that when specifically in "DEV_MODE", so i've added the NODE_ENV=development to the front of the npm install, so everything should work fine now.

lholmquist commented:
@bparees so if someone creates a new node project using `npm init` , a scripts property is added to the package.json,  but it could be possible that someone doesn't have it.

Although, in our run phase, we are running "npm run start"(by default), So a scripts property would be required or it would fail, so i think we might be ok with not checking for the existence of a scripts prop.  I'll create an issue anyway though, #153 

Thanks for the heads up on this

lholmquist commented:
So i got like 4 notifications for this, but i don't see the comment on GH



On Sun, Oct 21, 2018 at 7:43 PM Ben Parees <notifications@github.com> wrote:

> @lholmquist <https://github.com/lholmquist>
> fyi, this started breaking our tests in which we had an empty package.json
> ("{}"):
>
> [eval]:1
> console.log(require('./package.json').scripts.build ? true : false)
>                                              ^
>
> TypeError: Cannot read property 'build' of undefined
>     at [eval]:1:46
>     at Script.runInThisContext (vm.js:96:20)
>     at Object.runInThisContext (vm.js:303:38)
>     at Object.<anonymous> ([eval]-wrapper:6:22)
>     at Module._compile (internal/modules/cjs/loader.js:688:30)
>     at evalScript (internal/bootstrap/node.js:582:27)
>     at startup (internal/bootstrap/node.js:267:9)
>     at bootstrapNodeJSCore (internal/bootstrap/node.js:739:3)
> subprocess exited with status 1
> subprocess exited with status 1
>
> I don't know if that is considered a valid package.json or not, so up to
> you whether to do something about it, but basically anyone w/ no "scripts"
> entry is going to hit this i think.
>
if someone creates the package.json using "npm init", a script tag is
created, but i think this is something we should probably be  taking into
account,  someone not having a scripts tag.  Thanks for the heads up

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bucharest-gold/centos7-s2i-nodejs/pull/150#issuecomment-431714467>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA_TUssoeodJIWmOaBjmTEpyabCP7YJTks5unQaggaJpZM4XkJgk>
> .
>
